File: b3051.py
def main():
    n, k = tuple(int(i) for i in input().split())
    seq = [int(item) for item in input().split()]

    prefix_sum = {0: 1}

    qtd = 0
    curr_sum = 0

    for item in seq:
        curr_sum += item

        sum_before = curr_sum - k  # sum_before is if the sum of the range equals to k
        if sum_before in prefix_sum.keys():
            qtd += prefix_sum[sum_before]

        prefix_sum[curr_sum] = prefix_sum[curr_sum] + 1 if curr_sum in prefix_sum.keys() else 1

    print(qtd)


if __name__ == '__main__':
    main()

File: b3051.cpp
// https://www.beecrowd.com.br/judge/pt/problems/view/3051

#include <bits/stdc++.h>

using namespace std;

#define _ std::ios::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
#define dbg(x) cout << #x << " = " << x << endl
#define endl "\n"

typedef long long ll;

int main() { _
    ll n, k; cin >> n >> k;
    vector<ll> seq(n);

    for (auto& item : seq)
        cin >> item;

    // uma soma de prefixo da soma para a quantidade de vezes que essa soma aparece
    unordered_map<ll, ll> prefixSum {{0, 1}};

    ll qtd {0};
    ll currSum {0};

    for (ll i {0}; i < n; ++i) {
        currSum += seq[i];
        // se currSum - k está em prefixSum a soma entre a posição atual e a primeira quando currSum era igual à
        // currSum - k é igual à k
        if (prefixSum.find(currSum - k) != prefixSum.end())
            qtd += prefixSum[currSum - k];
        if (prefixSum.find(currSum) != prefixSum.end())
            prefixSum[currSum]++;
        else
            prefixSum[currSum] = 1;
    }

    cout << qtd << endl;
    
    return 0;
}
